### PR TITLE
(SERVER-852) Partial r10k support

### PR DESCRIPTION
**Pull request by nwolfe**
This commit adds the beginnings of configuring the master via an r10k
control repository. It goes so far as to run "r10k deploy" to map the
repository branches to Puppet environment directories, but does not do
anything after that like wire up any hiera.yaml configuration.

### CODE REVIEW

File: jenkins-integration/beaker/install/shared/r10k_deploy.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+end
+
+def run_r10k_deploy(host)
+  r10k = '/opt/puppetlabs/puppet/bin/r10k'
+  on(host, "#{r10k} deploy environment --puppetfile --verbose")
+end
+
+scenario_id = ENV['PUPPET_GATLING_SCENARIO']
+r10k_config = parse_scenario_file(scenario_id)['r10k']
+if not r10k_config.nil?
```
Comment (haus) [inline jenkins-integration/beaker/install/shared/r10k_deploy.rb]: `if r10k_config` is probably more idiomatic ruby

File: jenkins-integration/lib/puppet/gatling/config.rb [hunk 2]
```diff
@@ -65,3 +66,8 @@ def hiera_datadirs(hiera)
     [localpath, datadir]
   end
 end
+
+# Returns the path to the local r10k YAML configuration file
+def r10k_configpath(r10k)
+  File.join('config', 'r10ks', r10k + '.yaml')
```
Comment (haus) [inline jenkins-integration/lib/puppet/gatling/config.rb]: `r10ks` => `r10k`?

File: jenkins-integration/beaker/install/shared/r10k_deploy.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  scp_to(host, configfile, "#{configdir}/r10k.yaml")
+end
+
+def run_r10k_deploy(host)
+  r10k = '/opt/puppetlabs/puppet/bin/r10k'
+  on(host, "#{r10k} deploy environment --puppetfile --verbose")
+end
+
+scenario_id = ENV['PUPPET_GATLING_SCENARIO']
+r10k_config = parse_scenario_file(scenario_id)['r10k']
```
Comment (haus) [inline jenkins-integration/beaker/install/shared/r10k_deploy.rb]: It looks like parse_scenario_file will blow up if scenario_id is nil. That might be worth checking before calling the method.

Comment (nwolfe) [inline jenkins-integration/beaker/install/shared/r10k_deploy.rb]: Yep, that's true in every single one of these install/ scripts.  There's room for improvement certainly.  There isn't that much guarding in these scripts because ultimately they'll be called by a higher-level script/Jenkins job where we'll probably do more of that.

File: jenkins-integration/lib/puppet/gatling/config.rb (hunk 2 above)
Comment (nwolfe) [inline jenkins-integration/lib/puppet/gatling/config.rb]: `r10ks` is a directory full of r10k configuration files, following the existing pattern of `scenarios/`, `nodes/`, and `hieras/`

File: jenkins-integration/beaker/install/shared/r10k_deploy.rb (hunk 1 above)
Comment (nwolfe) [inline jenkins-integration/beaker/install/shared/r10k_deploy.rb]: Ah yes.

File: jenkins-integration/lib/puppet/gatling/config.rb (hunk 2 above)
Comment (haus) [inline jenkins-integration/lib/puppet/gatling/config.rb]: ah, cool

**Comment by nwolfe** [comment]
Yeah at some point I need to document how the hiera and r10k support fits together between the JSON files and the `hieras/` and `r10ks/` directories.

So far I've just been including examples of each but it's not exactly obvious that arbitrary strings need to match filenames under specific directories.

File: jenkins-integration/scripts/foss_install.sh [hunk 4]
```diff
@@ -1,6 +1,22 @@
...
+##
+## 2. Version of puppet-server package to install
+##    Can be specified with either:
+##    - ENV['PUPPETSERVER_BUILD_VERSION']=2.2.1
+##    - hosts.yaml#CONFIG#puppetserver_version=1.3.0
+
+if [ -z $BEAKER_CONFIG ]; then
+    echo "BEAKER_CONFIG hosts configuration required"
+    exit 1
+fi
```
Comment (cprice404) [inline jenkins-integration/scripts/foss_install.sh]: This doesn't seem related to the r10k stuff?

**Comment by cprice404** [comment]
@nwolfe I know this isn't complete, but it seems like it's basically a no-op unless someone puts in r10k stuff, so it seems like maybe we should just go ahead and merge it?  Maybe we should add a comment / log message somewhere that explains that it's not fully functional yet?

File: jenkins-integration/scripts/foss_install.sh (hunk 4 above)
Comment (nwolfe) [inline jenkins-integration/scripts/foss_install.sh]: It's not, but during development I run this file directly, and having this fail-fast is useful.

**Comment by nwolfe** [comment]
> I know this isn't complete, but it seems like it's basically a no-op unless someone puts in r10k stuff, so it seems like maybe we should just go ahead and merge it?

Yep, that's what I was thinking.

The `r10k_deploy.rb` file is for the most part complete (at least in terms of handling the r10k-specifics), it's just not wired up in the foss/pe_install files yet.

There's still the outstanding question of how to integrate hiera data from the r10k control repository, which we'll be avoiding for the specific OPS deployment scenario.

I'll add some comments about r10k and address other feedback.

**Comment by nwolfe** [comment]
Feedback addressed.  Ready for merge.
